wagtail/wagtailadmin/rich_text.py

```diff
@@ -44,15 +49,17 @@ def render(self, name, value, attrs=None):
 
     def render_js_init(self, id_, name, value):
         if self.options is not None and 'plugins' in self.options:
+            # explicit 'plugins' config passed in options, so use that
             plugin_data = self.options['plugins']
-        elif self.plugins is not None:
+        elif self.use_legacy_plugin_config:
+            # RemovedInWagtail114Warning
```

Comment:
Sorry, I really should have written something up to explain the relation between default features and `halloPlugins`, as the details are a bit convoluted...

TLDR: the plugin config obtained through the default features should always match `halloPlugins`, _on a vanilla Wagtail installation_. They will differ on existing installations that have added their own hallo.js customisations, and this setup is needed purely to accommodate them.

Currently, the official way to extend hallo.js is to edit the `halloPlugins` list. (More precisely, the official documented method is to call `registerHalloPlugin`, but that's much the same thing, and I know there are Wagtail installations in the wild where people have edited `halloPlugins` directly to _remove_ features.) We want to deprecate this and get people to use features instead - and we want to do so via our usual two-release deprecation policy.

The end goal is that we ditch the `halloPlugins` list (aka "the legacy plugin config"), and always use the config we obtain through calling `construct_plugins_list` on the feature list. However, we can't make that change yet, because then people would immediately lose the customisations they've made through the legacy plugin config. So, the interim solution I've come up with for Wagtail 1.12 - 1.13 is:

* If the rich text field has been passed an explicit `features` parameter, do everything the new way (since we can reasonably assume that a user who's actively using the new features functionality has ported all of the hallo customisations they care about into features)
* If no `features` parameter has been passed, use the legacy `halloPlugins` config; this ensures that any customisations that only exist through `halloPlugins` will remain in place.
* This means that our on-by-default features need to provide the new-style `construct_plugins_list` plumbing, and _also_ add themselves to the legacy `halloPlugins` config. In Wagtail 1.14, we will be able to drop the latter. (Third-party packages that provide enabled-by-default hallo plugins can drop this when they drop Wagtail <1.14 support.)

I don't think we can generate a deprecation warning for this, because the backwards-compatibility code path gets triggered any time the `features` param is omitted from the `RichTextField` (which is, and will continue to be, a valid thing to do). The exact deprecated behaviour we'd want to catch and warn against is "modifying `halloPlugins` without providing a corresponding feature definition", and the only way to spot that would be to somehow police what the Javascript code within `insert_editor_js` hooks does.